### ISSUE TITLE
Use .NET EventSource-compatible GUID if no GUID specified

### ISSUE THREAD

**Issue by zanedp** [microsoft/rust_win_etw#19]
When `trace_logging_provider` macro is not provided with a GUID, a .NET `EventSource`-compatible name-based GUID should be used.

Using the `uuid` crate with feature `v5`, this code will generate .NET `EventSource`-compatible GUIDs from the provider name:

```rust
use uuid::{uuid, Uuid};

const ETW_EVENT_SOURCE_NAMESPACE: Uuid = uuid!("482c2db2-c390-47c8-87f8-1a15bfc130fb");

fn etw_event_source_guid(provider_name: &str) -> Uuid {
    use sha1_smol::Sha1;

    let provider_bytes: Vec<u8> = provider_name
        .to_uppercase()
        .encode_utf16()
        .flat_map(|x| x.to_be_bytes())
        .collect();

    let mut hasher = Sha1::new();
    hasher.update(ETW_EVENT_SOURCE_NAMESPACE.as_bytes());
    hasher.update(&provider_bytes);

    let mut bytes = [0; 16];
    bytes.copy_from_slice(&hasher.digest().bytes()[..16]);

    // .NET EventSource reads the bytes from the SHA-1 hash as little-endian,
    // but does not set the "variant (aka type)" to the "Microsoft" value which
    // indicates little-endian in a RFC4122-conforming UUID.  The "variant" field
    // ends up being the "random" value copied from the from the SHA-1 hash, instead.
    uuid::Builder::from_bytes_le(bytes)
        .with_version(uuid::Version::Sha1)
        .into_uuid()
}
```

Some unit tests:

```rust
    assert_eq!(
        uuid!("5fefebda-b28e-5a81-d371-cebf3d3ddb41"),
        etw_event_source_guid("YourProviderName")
    );
    assert_eq!(
        uuid!("6f8eac67-f87f-598a-71a0-67e48d8c468d"),
        etw_event_source_guid("YourProviderNameYourProviderName")
    );
    assert_eq!(
        uuid!("5b1bddda-c110-53c9-b030-17aea3bfabd9"),
        etw_event_source_guid("YourProviderNameYourProviderNameYourProviderNameYourProviderName")
    );
    assert_eq!(
        uuid!("0cec4c9d-caa7-5d85-b3bd-c73577f03fd8"),
        etw_event_source_guid("Your.Provider.Name")
    );
    assert_eq!(
        uuid!("32805bf5-e0cc-522a-5f56-20c3e359ed93"),
        etw_event_source_guid("Your.Provider.Name.Your.Provider.Name")
    );
    assert_eq!(
        uuid!("ce5fa4ea-ab00-5402-8b76-9f76ac858fb5"),
        etw_event_source_guid("MyCompany.MyComponent")
    );
```

**Comment by zanedp** [comment]
Implemented in PR #20

**Comment by sivadeilra** [comment]
Thanks again!
